### PR TITLE
[RDY]:hi Normal works with -u NONE

### PR DESCRIPTION
**Pull request by teto**
The PR has 2 commits, the first one makes `:hi Normal works with -u NONE` works, just like vim.
This first commit also:
 - Makes HL_TABLE and ATTR_ENTTRY a function instead of a macro so that in can be used in gdb.
 - Introduces ATTRENTRY_INIT to init attrentry_tchanges, they make sense in the context of https://github.com/neovim/neovim/pull/6566 ; the long term goal is to be able to distinguish between NONE and an unset color.

The 2nd commit is just some documentation/styling changes (brackets) I had made to fix linting issues in another PR. I salvaged them but they are not necessary (might need to do another pass at it) so I can ditch it or send it another PR it that.

Nb: https://github.com/neovim/neovim/pull/6566 had become a mess so I am trying to split it into meaningful commits.

### CODE REVIEW

File: src/nvim/syntax.c [hunk 1]
```diff
@@ -7186,20 +7215,27 @@ int syn_namen2id(char_u *linep, int len)
 int syn_check_group(char_u *pp, int len)
 {
   char_u  *name = vim_strnsave(pp, len);
+
+  // HACK we need to make sure that the first highlight is the "Normal" hl
+  if (highlight_ga.ga_len == 0) {
+    int res = syn_add_group((char_u *)"Normal");
```
Comment (jamessan) [inline src/nvim/syntax.c]: `syn_add_group()` requires its argument to be an allocated string.  You're passing a static string instead.  Try this:

```c
int res = syn_add_group((char_u *)xstrdup("Normal"));
```

File: src/nvim/syntax_defs.h [hunk 2]
```diff
@@ -73,4 +73,14 @@ typedef struct attr_entry {
   int cterm_fg_color, cterm_bg_color;
 } attrentry_T;
 
+#define ATTRENTRY_INIT {.rgb_ae_attr = 0, .cterm_ae_attr = 0, \
```
Comment (jamessan) [inline src/nvim/syntax_defs.h]: Why not put `rgb_ae_attr` and `cterm_ae_attr` on their own lines too?  It looks a little odd having these two here but everything else on its own line.

File: src/nvim/syntax.c [hunk 3]
```diff
@@ -7186,20 +7217,27 @@ int syn_namen2id(char_u *linep, int len)
 int syn_check_group(char_u *pp, int len)
 {
   char_u  *name = vim_strnsave(pp, len);
+
+  // HACK we need to make sure that the first highlight is the "Normal" hl
+  if (highlight_ga.ga_len == 0) {
+    int res = syn_add_group((char_u *)xstrdup("Normal"));
+    assert(res == 0);
```
Comment (jamessan) [inline src/nvim/syntax.c]: Need to add a `(void)res;` line so release builds think `res` is used.

Comment (teto) [inline src/nvim/syntax.c]: Indeed. I would like to move the Normal hl creation to init_highlight but that would need more thoughts/tests as I am afraid ` :high Normal NONE`  or some other command could break things.

**Comment by jamessan** [comment]
```
From test_signs.vim:
Found errors in Test_sign():
function RunTheTest[9]..Test_sign line 19: Pattern '^sign Sign1 text=x sign Sign2 icon=../../pixmaps/stock_vim_find_help.png .*text=xy linehl=Error texthl=Title$' does not match 'sign Sign1 text=x sign Sign2 icon=../../pixmaps/stock_vim_find_help.png (not supported) text=xy linehl=MatchParen texthl=VertSplit'
```

File: src/nvim/syntax_defs.h (hunk 2 above)
Comment (teto) [inline src/nvim/syntax_defs.h]: done

**Comment by teto** [comment]
not sure https://travis-ci.org/neovim/neovim/jobs/251093995#L2732 is related  to my changes:
```
 [----------] Running tests from /home/travis/build/neovim/neovim/test/functional/autocmd/termclose_spec.lua

[ RUN      ] TermClose event triggers when terminal job ends: -- Output to stderr:

Vim: Caught deadly signal 'SIGHUP'
```
is it possible to restart the failed tests please ?

**Comment by jamessan** [comment]
> is it possible to restart the failed tests please ?

Done.

**Comment by teto** [comment]
I updated the first post to better reflect the current content.

**Comment by teto** [comment]
Would be happy to see this merged (even if only the first commit) as it solves the `:hi Normal`problem and helps me fixing https://github.com/neovim/neovim/pull/7082

File: src/nvim/syntax.c [hunk 4]
```diff
@@ -6270,12 +6270,12 @@ do_highlight(
...
-  /* Return if "default" was used and the group already has settings. */
-  if (dodefault && hl_has_settings(idx, TRUE))
+  // Return if "default" was used and the group already has settings
+  if (dodefault && hl_has_settings(idx, true)) {
     return;
+  }
 
-  if (STRCMP(HL_TABLE()[idx].sg_name_u, "NORMAL") == 0)
-    is_normal_group = TRUE;
+  is_normal_group = (STRCMP(HL_TABLE()[idx].sg_name_u, "NORMAL") == 0);
```
Comment (justinmk) [inline src/nvim/syntax.c]: before, `is_normal_group` kept its value if the if-block was not entered. Is this change intentional?

File: src/nvim/syntax.c [hunk 5]
```diff
@@ -6707,14 +6711,16 @@ static void highlight_clear(int idx)
...
 
-/*
- * Table with the specifications for an attribute number.
- * Note that this table is used by ALL buffers.  This is required because the
- * GUI can redraw at any time for any buffer.
- */
+/// Table with the specifications for an attribute number.
+/// Note that this table is used by ALL buffers.  This is required because the
+/// GUI can redraw at any time for any buffer.
+/// The attributes table
```
Comment (justinmk) [inline src/nvim/syntax.c]: the last line seems a bit unnecessary

File: src/nvim/syntax.c [hunk 6]
```diff
@@ -7124,10 +7143,10 @@ set_hl_attr (
   }
 }
 
-/*
- * Lookup a highlight group name and return it's ID.
- * If it is not found, 0 is returned.
- */
+/// Lookup a highlight group name and return it's ID.
```
Comment (justinmk) [inline src/nvim/syntax.c]: "its"

File: src/nvim/syntax.c [hunk 7]
```diff
@@ -7124,10 +7143,10 @@ set_hl_attr (
...
+///
+/// @param highlight name e.g. 'Cursor', 'Normal'
+/// @return the highlight id, else 0 if it \p name does not exist
```
Comment (justinmk) [inline src/nvim/syntax.c]: omit "it"

File: src/nvim/syntax.c [hunk 8]
```diff
@@ -7195,11 +7214,11 @@ int syn_check_group(char_u *pp, int len)
   return id;
 }
 
-/*
- * Add new highlight group and return it's ID.
- * "name" must be an allocated string, it will be consumed.
- * Return 0 for failure.
- */
+/// Add new highlight group and return it's ID.
```
Comment (justinmk) [inline src/nvim/syntax.c]: "its"

**Comment by justinmk** [comment]
Great to see smaller PRs, it helps a lot :>

LGTM, one comment: https://github.com/neovim/neovim/pull/6973#discussion_r134349010

File: src/nvim/syntax.c (hunk 4 above)
Comment (teto) [inline src/nvim/syntax.c]: it's the only place that sets is_normal_group so it's equivalent.

**Review by justinmk — APPROVED** [review]

**Comment by teto** [comment]
The failure does not seem related to the changes:
```
[----------] Running tests from /home/travis/build/neovim/neovim/test/functional/autocmd/termclose_spec.lua

[ RUN      ] TermClose event triggers when fast-exiting terminal job stops: -- Output to stderr:

Vim: Caught deadly signal 'SIGHUP'

Vim: Finished.

CMake Error at /home/travis/build/neovim/neovim/cmake/RunTests.cmake:53 (message):

  Running functional tests failed with error: 1.
```

**Comment by teto** [comment]
Thanks for the review, I rebased and pushed your fixes. Should be ok now.
